gto/git_utils.py

```diff
@@ -51,6 +52,80 @@ def wrapped_f(*args, **kwargs):
     return wrapped_f
 
 
+def commit_produced_changes_on_auto_commit(
+    message_generator: Callable[..., str] = generate_empty_commit_message
+):
+    """
+    The function `message_generator` can use any argument that the decorated function has.
+
+    Example: here we are using the argument b of the function f to generate the commit message
+
+        def create_message(b: str) -> str:
+            return "commit message with b={b}"
+
+        @commit_produced_changes_on_auto_commit(message_generator=create_message)
+        def f(a: str, b: str, c: str):
+            ...
+
+    """
+
+    def generate_commit_message(**kwargs) -> str:
+        kwargs_for_message_generator = {
+            k: kwargs[k] for k in inspect.getfullargspec(message_generator).args
+        }
+        return message_generator(**kwargs_for_message_generator)
+
+    def wrap(f: Callable):
+        @wraps(f)
+        def wrapped_f(*args, **kwargs):
+            kwargs = _turn_args_into_kwargs(f, args, kwargs)
+
+            if kwargs.get("auto_commit", False) is True:
+                if "repo" in kwargs:
+                    with stashed_changes(
+                        repo_path=kwargs["repo"], include_untracked=True
+                    ) as (stashed_tracked, stashed_untracked):
+                        result = f(**kwargs)
+                        if are_files_in_repo_changed(
+                            repo_path=kwargs["repo"],
+                            files=stashed_tracked + stashed_untracked,
+                        ):
+                            _reset_repo_to_head(repo_path=kwargs["repo"])
+                            raise GTOException(
+                                msg="The command changed files that were not committed, automated committing is not possible. "
+                                "The files have been rolled back to their previous state.\n\n"
+                                "Suggested action: Commit the changes and re-run this command"
+                            )
+                        git_add_and_commit_all_changes(
+                            repo_path=kwargs["repo"],
+                            message=generate_commit_message(**kwargs),
+                        )
+                else:
+                    raise ValueError(
+                        "Function decorated with commit_produced_changes_on_auto_commit was called with "
+                        "`auto_commit=True` but `repo` was not provided."
+                        "Argument `repo` is necessary."
+                    )
```

Comment:
User won't see this, only the developer may see it - when he wraps a function without `repo` arg, right?

Reply:
correct!